src/Dfc.CourseDirectory.Functions/FixVenues/FixApprenticeshipVenueReferences.cs

```diff
@@ -0,0 +1,528 @@
+        {
+            // _logger.LogDebug($"{LogPrefix} Analysing apprenticeshipId {apprenticeship.Id} ...");
+
+            // Thread-safe cache of venues for a provider.
+            // See https://andrewlock.net/making-getoradd-on-concurrentdictionary-thread-safe-using-lazy/
+            var providerVenues = await _venueCache.GetOrAdd(apprenticeship.ProviderUKPRN,
+                new Lazy<Task<IReadOnlyCollection<Venue>>>(() => GetVenues(apprenticeship.ProviderUKPRN))).Value;
+
+            var correction = _venueAnalyser.AnalyseApprenticeship(apprenticeship, providerVenues);
+
+            // debug output useful for picking ones out to test fix-specific endpoints
+            // var fixable = correction.ApprenticeshipLocationVenueCorrections.Count(l => l.VenueCorrection != null);
+            // if (fixable > 0) _logger.LogDebug($"Apprenticeship {apprenticeship.Id}: {fixable} fixable locations");
+
+            return correction;
+        }
+
+        private async Task<IReadOnlyCollection<Venue>> GetVenues(int ukprn)
+        {
+            try
+            {
+                var venues = await _cosmosDbQueryDispatcher.ExecuteQuery(
+                    new GetVenuesByProvider {ProviderUkprn = ukprn});
+
+                return venues.Where(v => v.Status == (int)VenueStatus.Live).ToList();
+            }
+            catch (Exception exception)
+            {
+                var message = $"Error fetching venues for ukprn {ukprn}. {exception.Message}"; // repeat exception in message because it's not showing in console output
+                _logger.LogError($"{LogPrefix} {message}", exception);
+                throw new Exception(message, exception);
+            }
+        }
+
+        private async Task ApplyCorrection(ApprenticeshipVenueCorrection apprenticeshipVenueCorrection, CancellationToken cancellationToken)
+        {
+            if (!_venueCorrector.Apply(apprenticeshipVenueCorrection)) // mutates apprenticeship directly
+            {
+                return; // no changes to apply
+            }
+
+            var documentUri = UriFactory.CreateDocumentUri(
+                _configuration.DatabaseId,
+                _configuration.ApprenticeshipCollectionName,
+                apprenticeshipVenueCorrection.Apprenticeship.Id.ToString());
+
+            // Dropping to native cosmosdb libraries here rather than cqrs commands because
+            // this gives us a complete round-trip without either:
+            // * introducing another read operation which would be significant on 40k records in terms of time and RU limits, or
+            // * blowing a hole in the command/query api by just exposing ReplaceDocument capability
+
+            _logger.LogDebug($"{LogPrefix} Replacing apprenticeship document at uri {documentUri} with corrected copy...");
+            try
+            {
+                var response = await _documentClient.ReplaceDocumentAsync(documentUri, apprenticeshipVenueCorrection.Apprenticeship,
```

Comment:
I'm a little surprised this works without specifying a PartitionKey

Reply:
https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.requestoptions.partitionkey?view=azure-dotnet#remarks

> For create and update operations on documents, the partition key is optional. When absent, the client library will extract the partition key from the document before sending the request to the server.